### PR TITLE
Staking part3

### PR DESCRIPTION
**Pull request by chaosma**
- Add CreateValidator/EditValidator messages processing logic
- Modify Staking Tx CLI to allow flexible testing; 
- createValidator/editValidator messages can handle correctly and store/retrieve correctly
- API for currentValidators and candidateValidators

Missing:
- Validator fields validity checking
- missing some field updating for validators
- validator snapshot function
- delegation message handling logic
- remaining API

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -2292,47 +2303,59 @@ func (bc *BlockChain) WriteStakingValidator(v *staking.ValidatorWrapper) error {
...
+	decodePayload, err := staking.RLPDecodeStakeMsg(payload, tx.StakingType())
+	if err != nil {
+		return err
+	}
 	switch tx.StakingType() {
 	case staking.DirectiveCreateValidator:
-		createValidator := tx.StakingMessage().(staking.CreateValidator)
-		m, err := bc.ReadValidatorMap()
+		createValidator := decodePayload.(*staking.CreateValidator)
+		list, err := bc.ReadValidatorList()
```
Comment (rlan35) [inline core/blockchain.go]: Let's add all validators in a single method so we don't have to read ValidatorList once for a single tx? Please add TODO if you don't have time.

File: core/state_transition.go [hunk 2]
```diff
@@ -347,10 +349,16 @@ func (st *StateTransition) applyEditValidatorTx(ev *staking.EditValidator, block
...
 	wrapper := st.state.GetStakingInfo(ev.ValidatorAddress)
+
+	oldRate := wrapper.Validator.Rate
 	if err := staking.UpdateValidatorFromEditMsg(&wrapper.Validator, ev); err != nil {
 		return err
 	}
-	wrapper.Validator.UpdateHeight = blockNum
+	newRate := wrapper.Validator.Rate
+	// update the commision rate change height
+	if oldRate.IsNil() || (!newRate.IsNil() && !oldRate.Equal(newRate)) {
```
Comment (rlan35) [inline core/state_transition.go]: Is the check of MaxChangeRate implemented?

File: node/node_handler.go [hunk 3]
```diff
@@ -410,6 +410,23 @@ func (node *Node) PostConsensusProcessing(newBlock *types.Block, commitSigAndBit
 // AddNewBlock is usedd to add new block into the blockchain.
 func (node *Node) AddNewBlock(newBlock *types.Block) error {
 	_, err := node.Blockchain().InsertChain([]*types.Block{newBlock}, true /* verifyHeaders */)
+
+	// Debug only
```
Comment (rlan35) [inline node/node_handler.go]: please remove debug code.

File: staking/types/validator.go [hunk 4]
```diff
@@ -27,10 +28,10 @@ var (
 
 // ValidatorWrapper contains validator and its delegation information
 type ValidatorWrapper struct {
-	Validator           `json:"validator" yaml:"validator"`
-	Delegations         []Delegation `json:"delegations" yaml:"delegations"`
-	SnapshotValidator   *Validator   `json:"snapshot_validator" yaml:"snaphost_validator"`
-	SnapshotDelegations []Delegation `json:"snapshot_delegations" yaml:"snapshot_delegations"`
+	Validator           `json:"validator" yaml:"validator" rlp:"nil"`
+	Delegations         []Delegation `json:"delegations" yaml:"delegations" rlp:"nil"`
+	SnapshotValidator   *Validator   `json:"snapshot_validator" yaml:"snaphost_validator" rlp:"nil"`
```
Comment (rlan35) [inline staking/types/validator.go]: only the *Validator can be nil right?    Others don't have to have "nil" tag. See example:

AccountNonce hexutil.Uint64  `json:"nonce"      gencodec:"required"`
		Price        *hexutil.Big    `json:"gasPrice"   gencodec:"required"`
		GasLimit     hexutil.Uint64  `json:"gas"        gencodec:"required"`
		ShardID      uint32          `json:"shardID"    gencodec:"required"`
		ToShardID    uint32          `json:"toShardID"`
		Recipient    *common.Address `json:"to"         rlp:"nil"`
		Amount       *hexutil.Big    `json:"value"      gencodec:"required"`
		Payload      hexutil.Bytes   `json:"input"      gencodec:"required"`
		V            *hexutil.Big    `json:"v" gencodec:"required"`
		R            *hexutil.Big    `json:"r" gencodec:"required"`
		S            *hexutil.Big    `json:"s" gencodec:"required"`
		Hash         *common.Hash    `json:"hash" rlp:"-"`

File: core/state_transition.go (hunk 2 above)
Comment (chaosma) [reply core/state_transition.go]: Not implemented, see the validity check TODO

File: node/node_handler.go (hunk 3 above)
Comment (chaosma) [reply node/node_handler.go]: I will comment it out. It will be useful later

File: staking/types/validator.go (hunk 4 above)
Comment (chaosma) [reply staking/types/validator.go]: Delegation can be nil as well if no body delegate. I saw similar decoding issue as well. Also, SnapshotValidator is not implemented and hence is nil. Thus, to be safe, I added nil to all places. Later, we can check one by one, until we are sure it's not nil.

**Pull request merged**